File: src/cmdbox/runtime/shell.py
import os
import sys
from shutil import which


def build_shell_command(command: str, preferred_shell: str | None = None) -> list[str]:
    """
    Builds a shell command list based on the provided command and preferred shell, with fallback mechanisms for
    platform compatibility and environment-specific configurations.

    This function determines the appropriate shell and constructs a command list for execution. It handles both
    Windows and Unix-like environments, prioritizing the specified preferred shell or defaulting to environment
    variables and standard fallback options. If no suitable shell is found, it raises a RuntimeError.

    Args:
        command (str): The shell command to execute.
        preferred_shell (str | None): The preferred shell to use for executing the command. If None, the function
            attempts to determine an appropriate shell based on the platform, environment, and fallback defaults.

    Returns:
        list[str]: A list containing the shell command and its arguments.

    Raises:
        RuntimeError: If no usable shell is found for the current system.
    """
    candidates: list[tuple[str, list[str]]] = []

    # Windows options
    if sys.platform.startswith("win"):
        if preferred_shell:
            candidates.append(
                (preferred_shell, _windows_shell_args(preferred_shell, command))
            )

        env_shell = os.environ.get("CMDBOX_SHELL")
        if env_shell:
            candidates.append((env_shell, _windows_shell_args(env_shell, command)))

        # Known good fallbacks
        candidates.extend(
            [
                ("pwsh", ["pwsh", "-NoProfile", "-Command", command]),
                ("powershell", ["powershell", "-NoProfile", "-Command", command]),
                ("cmd.exe", ["cmd.exe", "/C", command]),
            ]
        )

        for exe, args in candidates:
            if which(exe):
                return args

        # Last resort option
        return ["cmd.exe", "/C", command]

    # Unix options
    if preferred_shell:
        candidates.append((preferred_shell, [preferred_shell, "-lc", command]))

    env_shell = os.environ.get("SHELL")
    if env_shell:
        candidates.append((env_shell, [env_shell, "-lc", command]))

    candidates.extend(
        [
            ("/bin/bash", ["/bin/bash", "-lc", command]),
            ("/bin/sh", ["/bin/sh", "-lc", command]),
        ]
    )

    for exe, args in candidates:
        if os.path.isabs(exe):
            if os.path.exists(exe):
                return args
        elif which(exe):
            return args

    raise RuntimeError("No usable shell found for this system")


def _windows_shell_args(shell: str, command: str) -> list[str]:
    shell = shell.lower()

    if shell in ("pwsh", "powershell"):
        return [shell, "-NoProfile", "-Command", command]

    if shell in ("cmd", "cmd.exe"):
        return ["cmd.exe", "/C", command]

    # Treat unknown shell as executable
    return [shell, command]

File: internal/runtime/shell.go
package runtime

import (
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
)

type candidate struct {
	exe  string
	args []string
}

// BuildShellCommand constructs a shell command for the given platform, using the preferred or available shell executables.
func BuildShellCommand(command string, preferredShell string) ([]string, error) {
	var candidates []candidate

	// Windows options
	if runtime.GOOS == "windows" {
		if preferredShell != "" {
			candidates = append(candidates, candidate{
				exe:  preferredShell,
				args: windowsShellArgs(preferredShell, command),
			})
		}

		envShell, exists := os.LookupEnv("CMDBOX_SHELL")
		if exists {
			candidates = append(candidates, candidate{
				exe:  envShell,
				args: windowsShellArgs(envShell, command),
			})
		}

		// Known good fallbacks
		candidates = append(candidates,
			candidate{exe: "pwsh", args: []string{"pwsh", "-NoProfile", "-Command", command}},
			candidate{exe: "powershell", args: []string{"powershell", "-NoProfile", "-Command", command}},
			candidate{exe: "cmd.exe", args: []string{"cmd.exe", "/C", command}},
		)

		for _, c := range candidates {
			if which(c.exe) {
				return c.args, nil
			}
		}

		// Last resort option
		return []string{"cmd.exe", "/C", command}, nil
	}

	if preferredShell != "" {
		candidates = append(candidates, candidate{
			exe:  preferredShell,
			args: []string{preferredShell, "-lc", command},
		})
	}

	if envShell := os.Getenv("SHELL"); envShell != "" {
		candidates = append(candidates, candidate{
			exe:  envShell,
			args: []string{envShell, "-lc", command},
		})
	}

	candidates = append(candidates,
		candidate{exe: "/bin/bash", args: []string{"/bin/bash", "-lc", command}},
		candidate{exe: "/bin/sh", args: []string{"/bin/sh", "-c", command}},
	)

	for _, c := range candidates {
		if filepath.IsAbs(c.exe) {
			if _, err := os.Stat(c.exe); err == nil {
				return c.args, nil
			}
		} else if which(c.exe) {
			return c.args, nil
		}
	}

	return nil, ErrNoUsableShell
}

// windowsShellArgs returns the argument list required to execute a command in the specified Windows shell.
func windowsShellArgs(shell string, command string) []string {
	shell = strings.ToLower(shell)
	if shell == "pwsh" || shell == "powershell" {
		return []string{shell, "-NoProfile", "-Command", command}
	}
	if shell == "cmd" || shell == "cmd.exe" {
		return []string{"cmd.exe", "/C", command}
	}
	return []string{shell, command} // Treat unknown shell as executable
}

// which checks if the given executable can be found, either as an absolute/relative
// path that exists or by searching PATH for its name.
func which(exe string) bool {
	_, err := exec.LookPath(exe)
	return err == nil
}
